src/main/java/carpet/api/settings/SettingsManager.java

```diff
@@ -0,0 +1,781 @@
+import carpet.settings.ParsedRule;
+import carpet.settings.Rule;
+import carpet.utils.CommandHelper;
+import carpet.utils.Messenger;
+import carpet.utils.TranslationKeys;
+import carpet.utils.Translations;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.loader.api.FabricLoader;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.SharedSuggestionProvider;
+import net.minecraft.network.chat.BaseComponent;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.level.storage.LevelResource;
+
+import static carpet.script.CarpetEventServer.Event.CARPET_RULE_CHANGES;
+import static carpet.utils.Translations.tr;
+import static net.minecraft.commands.Commands.argument;
+import static net.minecraft.commands.Commands.literal;
+import static net.minecraft.commands.SharedSuggestionProvider.suggest;
+
+/**
+ * <p>A {@link SettingsManager} is a class that manages {@link CarpetRule carpet rules} in a {@link MinecraftServer},
+ * including a command to manage it, and allowing other code to hook into rule changes by using {@link RuleObserver rule observers}.</p>
+ * 
+ * <p>You can create your own {@link SettingsManager} if you want to have your own command that will be handled the same way as Carpet's
+ * by using {@link #SettingsManager(String, String, String)} and returning it in your {@link CarpetExtension}'s
+ * {@link CarpetExtension#extensionSettingsManager()} method.</p>
+ */
+@SuppressWarnings({"deprecation", "removal"}) // remove after removing old system
+public class SettingsManager {
+    private final Map<String, CarpetRule<?>> rules = new HashMap<>();
+    private final String version;
+    private final String identifier;
+    private final String fancyName;
+    private boolean locked;
+    private MinecraftServer server;
+    private final List<RuleObserver> observers = new ArrayList<>();
+    private static final List<RuleObserver> staticObservers = new ArrayList<>();
+    static record ConfigFileData(Map<String, String> ruleMap, boolean locked, List<String> comments) {}
+    
+    /**
+     * <p>Defines a class that can be notified about a {@link CarpetRule} changing.</p>
+     * 
+     * @see #ruleChanged(CommandSourceStack, CarpetRule)
+     * @see SettingsManager#addRuleObserver(RuleObserver)
+     * @see SettingsManager#addGlobalRuleObserver(RuleObserver)
+     */
+    @FunctionalInterface
+    public static interface RuleObserver {
+        /**
+         * @param source The {@link CommandSourceStack} that likely originated this change, and should be the notified source for further
+         *               messages. Can be {@code null} if there was none and the operation shouldn't send feedback.
+         * @param changedRule The {@link CarpetRule} that changed. Use {@link CarpetRule#value() changedRule.value()} to get the rule's value,
+         *                    and pass it to {@link RuleHelper#toRuleString(Object)} to get the {@link String} version of it
```

Comment:
you have missed the point of the third parameter - the string value. Its not the machine translated and understood value set for this rule, but the actual string that player has typed.

I don't see a reason for removing it

Reply:
I wanted to keep it at the start, but ended up not doing so because a technical issue.

If you add the third parameter, then you have the original (fallback) `addRuleObserver(TriConsumer)` and the new `addRuleObserver(RuleObserver)`. While that's not a compile issue for the `SettingsManager` class as two methods can have the same name as long as they have different parameters, then consumers of the class can't call it with a lambda, as both the old and new have the same simple signature (3 params) and the method is ambiguous to the compiler. It may work if they specify the parameter types when declaring the lambda, but it makes it less easy to use, and would break source compatibility (existing binaries would be fine). Other alternatives were renaming the method, but I couldn't think of how to call it.

At the end of the day I thought it wasn't _that_ useful so I decided removing it was the easiest way (at the end of the day, this only fires if the rule was changed successfully, and then most of the time the value as a string will be the input).

How could I call the method then?

Reply:
Fixed by renaming to registerRuleObserver.